### ISSUE TITLE
Potential Unsound: 1 out-of-bound read and 5 unaligned memory access.

### ISSUE THREAD

**Issue by YoshikiTakashima** [hsivonen/encoding_rs#52]
Hello.

I'm Yoshiki, a PhD student at CMU. 

We are testing a tool to automatically generate test cases from API data and existing tests.

A few of our generated test cases were reported as "unsound" by Miri, mostly due to unaligned or out-of-bound memory. I've attached a [Tarball](https://github.com/hsivonen/encoding_rs/files/5066822/encoding_rs_miri_flagged.tar.gz) that contains the test cases that induce this behavior.

Please note that, because the framework leverages existing tests as templates, some of the test cases overlap with existing test cases for the library. In particular,
```
decode(BIG5, b"", &"");//LAYER:0
```
also shows up in the manually written [test cases](https://github.com/hsivonen/encoding_rs/blob/master/src/big5.rs#L282).

In case this is intended behavior, or you would prefer if I focused on other parts of the code, please let me know.

Thanks.
~Yoshiki

**Comment by hsivonen** [comment]
Thank you!

My results so far:

* I can't reproduce the `out_of_bound` case locally in Miri.
* I have stepped through cases 118, 140, and 596 in x86 debug executions, and the computations were OK.

Based on reading the error messages and based on my previous discussion with @RalfJung, I believe these to be instances of the [Miri issue](https://github.com/rust-lang/miri/issues/1074) mentioned in the error messages.

Due to predating `align_to` in the standard library, `encoding_rs` does its own alignment math. As I understand it, the Miri errors are based on pointers carrying alignment metadata in the Miri execution as opposed to Miri actually checking whether pointer math has been done correctly when casting from a pointer with smaller alignment to a pointer with a larger alignment.

Also, as I understand it, migrating to `align_to` would result in the interesting code paths getting tested but would result in Miri taking the least interesting code path.

Does this analysis look correct to you?

**Comment by RalfJung** [comment]
@hsivonen what you say makes sense for unaligned accesses, but should not lead to out-of-bounds errors.

I should really prioritize fixing that alignment problem...

**Comment by YoshikiTakashima** [comment]
Hi @hsivonen.

Thank you for the analysis on alignment. I agree with you that these are safe behaviors given the nature of the library.

With regard to out-of-bound, I should correct myself that what MIRI reports is not a read per-se, but rather creating a pointer that is outside the bounds of allocated memory. If you run the `cargo miri test` inside the attached tarball, you should get a long trace and an error that looks like this. 
```
error: Undefined Behavior: inbounds test failed: pointer must be in-bounds at offset 10, but is outside bounds of alloc93891 which has size 2
```
Is the `out_of_bound()`  test case passing on your side? 

Since `decode(BIG5, b"", &"");` also shows up in the manually written test cases, I was able to replicate the identical behavior by cloning the latest version of this repo, and running `cargo miri test` inside the cloned directory. Although, in this case, the memory offset will be 8 and you will have to wait about 15 seconds for MIRI to reach that specific test case.

For your information, my toolchain versions are:
```
$ cargo --version
cargo 1.46.0-nightly (fede83ccf 2020-07-02)
$ cargo miri --version
miri 0.1.0 (fd81012 2020-06-28)
```
and the tarball imports `encoding_rs = "0.8.23"`.

Finally, thank you for the help @RalfJung.

Thanks,
~Yoshiki

**Comment by hsivonen** [comment]
@RalfJung @YoshikiTakashima Thank you. The OOB case indeed materialized an OOB pointer though didn't dereference it. Fixed in #53 . I'm leaving this issue open to remind me to check the SIMD code for the same pattern.

**Comment by YoshikiTakashima** [comment]
@hsivonen Got it.

Big thanks to @bjorn3, who suggested the fix.

**Comment by hsivonen** [comment]
> @RalfJung @YoshikiTakashima Thank you. The OOB case indeed materialized an OOB pointer though didn't dereference it. Fixed in #53 . I'm leaving this issue open to remind me to check the SIMD code for the same pattern.

For clarity, closing this issue and spinning revision of the SIMD code into a [distinct issue](https://github.com/hsivonen/encoding_rs/issues/87).
